sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PushPredicateThroughJoin.scala

```diff
@@ -0,0 +1,248 @@
+      Seq.empty
+    }
+
+    // To avoid expanding the join condition into conjunctive normal form and making the size
+    // of codegen much larger, `commonCondition` will be kept as original form in the new join
+    // condition.
+    (leftEvaluateCondition ++ leftExtraCondition, rightEvaluateCondition ++ rightExtraCondition,
+      commonCondition ++ nonDeterministic)
+  }
+
+  private def convertibleFilter(
+    condition: Expression,
+    outputSet: AttributeSet,
+    canPartialPushDown: Boolean): Option[Expression] = condition match {
+    // At here, it is not safe to just convert one side and remove the other side
+    // if we do not understand what the parent filters are.
+    //
+    // Here is an example used to explain the reason.
+    // Let's say we have NOT(a = 2 AND b in ('1')) and we do not understand how to
+    // convert b in ('1'). If we only convert a = 2, we will end up with a filter
+    // NOT(a = 2), which will generate wrong results.
+    //
+    // Pushing one side of AND down is only safe to do at the top level or in the child
+    // AND before hitting NOT or OR conditions, and in this case, the unsupported predicate
+    // can be safely removed.
+    case And(left, right) =>
+      val leftResultOptional = convertibleFilter(left, outputSet, canPartialPushDown)
+      val rightResultOptional = convertibleFilter(right, outputSet, canPartialPushDown)
+      (leftResultOptional, rightResultOptional) match {
+        case (Some(leftResult), Some(rightResult)) => Some(And(leftResult, rightResult))
+        case (Some(leftResult), None) if canPartialPushDown => Some(leftResult)
+        case (None, Some(rightResult)) if canPartialPushDown => Some(rightResult)
+        case _ => None
+      }
+
+    // The Or predicate is convertible when both of its children can be pushed down.
+    // That is to say, if one/both of the children can be partially pushed down, the Or
+    // predicate can be partially pushed down as well.
+    //
+    // Here is an example used to explain the reason.
+    // Let's say we have
+    // (a1 AND a2) OR (b1 AND b2),
+    // a1 and b1 is convertible, while a2 and b2 is not.
+    // The predicate can be converted as
+    // (a1 OR b1) AND (a1 OR b2) AND (a2 OR b1) AND (a2 OR b2)
+    // As per the logical in And predicate, we can push down (a1 OR b1).
+    case Or(left, right) =>
+      for {
+        lhs <- convertibleFilter(left, outputSet, canPartialPushDown)
+        rhs <- convertibleFilter(right, outputSet, canPartialPushDown)
+      } yield Or(lhs, rhs)
+
+    case Not(pred) =>
+      val childResultOptional = convertibleFilter(pred, outputSet, canPartialPushDown = false)
+      childResultOptional.map(Not)
+
+    case other =>
+      if (other.references.subsetOf(outputSet)) {
+        Some(other)
+      } else {
+        None
+      }
+  }
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform applyLocally
+
+  val applyLocally: PartialFunction[LogicalPlan, LogicalPlan] = {
```

Comment:
This method is copied from the original one, except that the join condition is not changed when `enablePushingExtraPredicates` is true